R/detect.r

```diff
@@ -104,3 +104,53 @@ str_ends <- function(string, pattern, negate = FALSE) {
          }
   )
 }
+
+#' Pattern Conversion Between LIKE and Regex structures
+#'
+#' Turns a SQL LIKE structure into a recognisible regular expression pattern.
+#'
+#' @param pattern SQL LIKE operator pattern to be converted.
+#'
+#' @return A converted pattern.
+#'
+#' @examples
+#' like_converter("bana%")
+#' like_converter("app_e")
+like_converter <- function(pattern) {
+  converted <- stri_replace_all_regex(pattern, "(?<!\\\\|\\[)%(?!\\])", "\\.\\*")
+  converted <- stri_replace_all_regex(converted, "(?<!\\\\|\\[)_(?!\\])", "\\.")
+  paste0("^", converted, "$")
+}
+
+#' Detect the presence of a pattern in the string using SQL LIKE convention.
+#'
+#' Vectorised over `string` and `pattern`.
+#' Follows the structure of the SQL `LIKE` operator and allows the use of their
+#'  wild cards `%` or `_`.
+#' These can be escaped using `\\\\` or `[]`.
+#'
+#' @inheritParams str_detect
+#' @param ignore_case Define if the case should be ignored.
+#'  Defaults to `TRUE` to match the SQL `LIKE` operator.
+#'
+#' @return A logical vector.
+#' @export
+#'
+#' @examples
+#' fruit <- c("apple", "banana", "pear", "pinapple")
+#' str_like(fruit, "app%")
+#' str_like(fruit, "ba_ana")
+#' str_like(fruit, "%APPLE")
+str_like <- function(string, pattern, negate = FALSE, ignore_case = TRUE) {
+  switch(type(pattern),
```

Comment:
I think you could now simplify this to:

```
if (type(pattern) == "regex") {
  pattern <- like_converter(pattern)
}
str_detect(string, pattern, negate = negate, ignore_case = ignore_case)
```

Does that make sense?

Reply:
Hmmm, or should this function accept the other pattern types at all? Maybe that doesn't make sense for a more specialised function. In that case the function would look more like:

```R
switch(type(pattern), 
  regex = stri_detect_regex(string, like_converter(pattern), negate = negate, opts_regex = opts(pattern)),
  stop("str_like() works only with regular expressions", call. = FALSE
)
```

Reply:
I agree that this should probably just accept a regex, have updated similar to above and made a separate call in order to preserve the ignore case earlier.